plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -632,14 +634,71 @@ private CpsScript parseScript() throws IOException {
             }
-        } catch (RuntimeException | Error x) {
+        } catch (Exception x) {
+            // Suspected groovyjarjarasm.asm.MethodTooLargeException or a
+            // org.codehaus.groovy.control.MultipleCompilationErrorsException
+            // whose collection of errors refers to MethodTooLargeException.
+            // Per review comments, we do not want to statically compile a
+            // dependency on the groovyjarjarasm.asm.MethodTooLargeException
+            // internals, so gauge hitting it via String name comparisons.
+            // Other cases may be (subclasses of) RuntimeException or Error.
+            Exception mtlEx = null;
+            int ecCount = 0;
+
+            // Clean up first
             closeShells();
-            throw x;
+
+            if (x.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                mtlEx = x;
+                ecCount = 1;
+            } else if (x instanceof MultipleCompilationErrorsException) {
+                ErrorCollector ec = ((MultipleCompilationErrorsException) x).getErrorCollector();
+                ecCount = ec.getErrorCount();
+
+                for (int i = 0; i < ecCount; i++) {
+                    Exception ex = ec.getException(i);
+                    if (ex == null)
+                        continue;
+
+                    LOGGER.log(Level.FINE, "Collected Exception #" + i + ": " + ex.toString());
+                    if (ex.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                        mtlEx = ex;
+                        break;
+                    }
+                }
+            }
+
+            if (mtlEx == null || ecCount < 1) {
+                // Some other exception type, or collection did not include MTL, rethrow as-is
+                throw x;
+            }
+
+            String msg = "FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException";
+            if (ecCount > 1) {
+                msg += " (and other issues)";
+            }
+            // Short message suffices, not much that a pipeline developer
+            // can do with the stack trace into the guts of groovy
+            msg += "; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library: ";
+            msg += mtlEx.getMessage();
+
+            if (ecCount > 1) {
+                // Not squashing with explicit MethodTooLargeException
+                // re-thrown below, in this codepath we have other errors.
+                throw new RuntimeException(msg, x);
+            } else {
+                // Do not confuse pipeline devs by a wall of text in the
+                // build console, but let the full context be found in
+                // server log with some dedication.
+                LOGGER.log(Level.FINE, mtlEx.getMessage());
+                //throw new RuntimeException(msg, mtlEx);
+                throw new RuntimeException(msg);
+            }
```

Comment:
Regarding the concern of "pretty convoluted logic" - this block is largely why I did it so: groovy can return a `MultipleCompilationErrorsException` with an attached collection of one or more errors (broadly speaking), so we need to count if we *only* had the `MethodTooLargeException` one way or another here (and then print the pretty log), or also something else and then better print everything for human devs to sift through it diligently.